misc(generator): Allow local paths to generators

dylanonelson opened the pull request:
This is a first attempt at fixing #262. I'd like some feedback on this approach before I add tests and hunt down the corresponding documentation. Thanks!

**What kind of change does this PR introduce?**
feature

**Did you add tests for your changes?**
Not yet! Would like some feedback on this approach before I write tests.

**If relevant, did you update the documentation?**
I removed a line from `SCAFFOLDING.md`, but there are probably other places where the documentation needs updating.

**Summary**
Fixes #262. Allows users to pass a relative path to the `init` command so that they can compose generator packages not published to npm.

**Does this PR introduce a breaking change?**
No.

**Other information**
- The current implementation treats the package name as a potential relative path by default. It does not validate that the path starts with `./` or `../`.
- It assumes that the user‘s local generator packages are ready for `require`ing—i.e., that they have all their dependencies installed. It does not link them globally.

ev1stensberg commented:
Hi, thanks for the PR! Could you manage to fix the issue you posted as well in this PR?

dylanonelson commented:
Sure thing, will do. Will take me a few days to get to it.

ev1stensberg commented:
Gotcha, no worries!

dylanonelson commented:
I've updated the pr to fix #266 as well. I know I need to add tests, but since this is my first js pr I'd like feedback on the way it's implemented before I spend the time to set them up. There are some nuances to how the cli validates and resolves package names that you may have different ideas about. Thanks!

dylanonelson commented:
@ev1stensberg Any chance you or another maintainer would have time to take a look at this sometime soon?

ev1stensberg commented:
@dylanonelson I'll have a look at it later today 👍

ev1stensberg commented on lib/utils/package-manager.js:
```diff
@@ -71,7 +72,31 @@ function getPackageManager() {
+ *
+ * @returns {String} path - Path to global node_modules folder
+ */
+function getPathToGlobalPackages() {
```
Is this really needed? We already have a package manager function that takes care of this?

dylanonelson replied:
Which function are you referring to? I don't see any code in `pacakge-manager.js` that serves the same purpose.

ev1stensberg commented on lib/utils/npm-packages-exists.js:
```diff
@@ -1,6 +1,7 @@
 const chalk = require("chalk");
-const npmExists = require("./npm-exists");
+const packageExists = require("./package-exists");
+const path = require("path");
```
Path should be above packageExists

EugeneHlushko replied:
maybe its work enabling absolute imports eslint check?

ev1stensberg commented on SCAFFOLDING.md:
```diff
@@ -14,7 +14,7 @@ Before writing a `webpack-cli` scaffold, think about what you're trying to achie
 ## webpack-addons-yourpackage
 
-In order for `webpack-cli` to compile your package, it relies on a prefix of `webpack-addons`. The package must also be published on npm. If you are curious about how you can create your very own `addon`, please read [How do I compose a webpack-addon?](https://github.com/ev1stensberg/webpack-addons-demo).
+In order for `webpack-cli` to compile your package, it relies on a prefix of `webpack-addons`. If you are curious about how you can create your very own `addon`, please read [How do I compose a webpack-addon?](https://github.com/ev1stensberg/webpack-addons-demo).
```
Change this to describe diff from local packages ( no prefix needed ) and npm ones

dylanonelson replied:
Ah OK, I didn't realize that local packages didn't have to have the prefix.

ev1stensberg commented on lib/utils/package-manager.js:
```diff
@@ -48,7 +48,8 @@ function spawnYarn(pkg, isNew) {
 function spawnChild(pkg) {
-	const pkgPath = path.resolve(globalPath, pkg);
+	const rootPath = getPathToGlobalPackages();
+	const pkgPath = path.resolve(rootPath, pkg);
```
Okay lgtm for now 👍

EugeneHlushko commented on lib/utils/package-exists.js:
```diff
@@ -0,0 +1,22 @@
+ */
+
+module.exports = function packageExists(pkg) {
+	const isValidLocalPath = fs.existsSync(pkg);
```
not sure its worth caching into varaible as its a sync call

ev1stensberg commented:
Any update on this?

dylanonelson commented:
This week was busy for me, but I was just going to take another look today. Will push some changes by tomorrow.

ev1stensberg commented:
No worries, there's no rats race. Take your time 💙

dylanonelson commented:
I'm a little confused by this output because it says the tests are all passing. So where are the errors?

ev1stensberg commented:
It's your commit format, if you so a biopsy commit using git add followed up by git cz you can see how they're supposed to be

dylanonelson commented:
I think this is ready for another review, @EugeneHlushko @ev1stensberg.

I've added some tests for `npm-packages-exists.js` and for the function I added in `package-manager.js`. I ran into issues trying to test `resolve-packages.js`, because the code I added calls `require.resolve`, and I couldn't figure out a clean way to mock that function. Some of these tests seem very closely tied to the implementations they're meant to test, but perhaps that's inevitable given the number of IO operations in the code. Would appreciate any feedback on that topic.

I should add that the test file for `resolve-packages.js` is a bit strange, at least to my eyes—it doesn't actually import `resolve-packages.js`, so I'm not sure what it's testing.

EugeneHlushko commented on SCAFFOLDING.md:
```diff
@@ -14,7 +14,11 @@ Before writing a `webpack-cli` scaffold, think about what you're trying to achie
+
+If the package is on npm, its name must have a prefix of `webpack-addons`.
+
+If the package is on your local filesystem, it can be named whatever you want. Pass the name of the package as a relative path to its root directory.
```
i just noticed that you've removed `[How do I compose a webpack-addon?](https://github.com/ev1stensberg/webpack-addons-demo).` I would put it back was a good reference to me some time ago.

dylanonelson replied:
Good catch, I'll restore that.

EugeneHlushko commented on lib/utils/resolve-packages.js:
```diff
@@ -41,10 +43,33 @@ function resolvePackages(pkg) {
+				require.resolve(absolutePath);
+				packageLocations.push(absolutePath);
+			} catch (err) {
+				console.log(`Cannot find a valid npm module at ${absolutePath}.`);
```
Sorry if this was asked before, why we dont `throw` when provided module wasnt found?

dylanonelson replied:
Good point—I think that's just a mistake on my part. I will use the code in the `catch` below to log the error and exit the process if we can't resolve the package.

ev1stensberg commented:
@dylanonelson ping me once this is R4R

dylanonelson commented:
I think it is ready for review now, @ev1stensberg, unless you want me to make the two small changes Eugene requested first? I'll make those today.

ev1stensberg commented:
Yep, do that!

dylanonelson commented:
@ev1stensberg alright, I've updated the commit based on Eugene's feedback.

ev1stensberg commented:
Mind fixing conflicts and I'll re-review?

dylanonelson commented:
@ev1stensberg done.

ev1stensberg commented on lib/utils/npm-packages-exists.js:
```diff
@@ -14,8 +15,22 @@ const resolvePackages = require("./resolve-packages").resolvePackages;
+
+		// The addon is on npm; validate name and existence
+		// eslint-disable-next-line
 		if (addon.length <= 14 || addon.slice(0, 14) !== "webpack-addons") {
```
If you're passing the relative path you'll get `../webpack-cli/lib/generators/generator.js isn't a valid name.` and `It should be prefixed with 'webpack-addons', but have different suffix.`. Can you do so it says `../webpack-cli/lib/generators/generator.js isn't a valid name or path.` ? And we shouldn't throw the addons error if it's a path

dylanonelson replied:
I think the `return` statement on line 28 means you'll never make it to that error code.

ev1stensberg replied:
I got there :(
<img width="561" alt="skjermbilde 2018-03-03 kl 02 44 14" src="https://user-images.githubusercontent.com/16735925/36929098-cea0cb58-1e8c-11e8-91c7-a59c80d3da45.png">

dylanonelson replied:
Ah OK I see. I will fix it.

ev1stensberg replied:
Sweet!

dylanonelson replied:
Oh wait—it looks like you ran the command pointing at the generator js file. Maybe I misunderstood the requirements. Do we want to allow people to point the `init` command at a javascript file, an npm package, or both?

dylanonelson replied:
Ugh, I can't reproduce the error you're seeing. Not sure what the deal is. When I pass a relative path to a generator on my filesystem it resolves it without throwing the prefix error. I'm going to be out of town till next Wednesday, so I won't be able to investigate further till then.

ev1stensberg replied:
both 👍

dylanonelson commented:
@ev1stensberg had an hour tonight to spare. I think my changes address the error you were seeing. If you review in the next few days I can pick it up again next week.

dylanonelson commented on lib/utils/npm-packages-exists.js:
```diff
@@ -20,7 +20,10 @@ module.exports = function npmPackagesExists(pkg) {
 
 		//eslint-disable-next-line
-		if (addonBasename.length <= 14 || addonBasename.slice(0, 14) !== "webpack-addons") {
+		if (
```
For sure. Just pushed a commit to refactor that code.

fokusferit commented on lib/utils/package-manager.js:
```diff
@@ -71,7 +72,34 @@ function getPackageManager() {
+		}
+	}
+
+	return globalPath;
```
could we put the require of this into the function? So that in the future no one mixes maybe `globalPath` with `getPathToGlobalPackage()`. 

```
const globalPath = require("global-modules");
const manager = getPackageManager();
```
Then the scope of the function is completely independent.

dylanonelson replied:
Sure!

fokusferit commented on lib/utils/package-manager.spec.js:
```diff
@@ -65,7 +72,7 @@ describe("package-manager", () => {
 		const packageName = "some-pkg";
 
-		mockSpawnErrorOnce();
+		mockSpawnErrorTwice();
```
May I ask, what the reason for having same spawn calls twice instead of one ?

dylanonelson replied:
`spawnChild` now calls `spawn.sync` twice, because it calls `getPackageManager` twice. It calls `getPackageManager` once and also calls `getPathToGlobalPackages`. which in turn calls `getPackageManager`. I can refactor it if that seems too redundant.

ev1stensberg commented:
@dylanonelson @fokusferit left some comments, could you open a PR to fix those? Need your changes to another PR
